File: aegis/sdk/go/aegis/validation.go
package aegis

import "encoding/json"

type ValidationFailure struct {
	Code    ErrorCode `json:"code"`
	Path    string    `json:"path"`
	Message string    `json:"message"`
}

type ValidationResult struct {
	Outcome  ValidationOutcome  `json:"outcome"`
	Failures []ValidationFailure `json:"failures,omitempty"`
	Metadata map[string]string  `json:"metadata,omitempty"`
}

func (r ValidationResult) Valid() bool {
	return r.Outcome == OutcomeValid
}

func (r ValidationResult) JSON() ([]byte, error) {
	return json.Marshal(r)
}

func validResult(metadata map[string]string) ValidationResult {
	return ValidationResult{Outcome: OutcomeValid, Metadata: metadata}
}

func invalidResult(failures []ValidationFailure) ValidationResult {
	return ValidationResult{Outcome: OutcomeInvalid, Failures: failures}
}

func failure(code ErrorCode, path string, message string) ValidationFailure {
	return ValidationFailure{Code: code, Path: path, Message: message}
}

File: aegis/sdk/python/aegis/validation.py
from dataclasses import dataclass, field
from typing import Any

from .errors import ErrorCode, ValidationOutcome


@dataclass(frozen=True)
class ValidationFailure:
    code: ErrorCode
    path: str
    message: str


@dataclass(frozen=True)
class ValidationResult:
    outcome: ValidationOutcome
    failures: tuple[ValidationFailure, ...] = field(default_factory=tuple)
    metadata: dict[str, Any] = field(default_factory=dict)

    @property
    def valid(self) -> bool:
        return self.outcome == ValidationOutcome.VALID


def valid_result(**metadata: Any) -> ValidationResult:
    return ValidationResult(ValidationOutcome.VALID, metadata=metadata)


def invalid_result(failures: list[ValidationFailure]) -> ValidationResult:
    return ValidationResult(ValidationOutcome.INVALID, tuple(failures))


def failure(code: ErrorCode, path: str, message: str) -> ValidationFailure:
    return ValidationFailure(code=code, path=path, message=message)
